firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLitePersistence.java

```diff
@@ -143,35 +143,32 @@ RemoteDocumentCache getRemoteDocumentCache() {
 
   @Override
   void runTransaction(String action, Runnable operation) {
+    Logger.debug(TAG, "Starting transaction: %s", action);
+    db.beginTransactionWithListener(referenceDelegate);
```

Comment:
Since ReferenceDelegate has public onTransactionStarted() / onTransactionCommitted() methods as part of its interface contract, I think it might be cleaner to leave SQLiteLruReferenceDelegate alone, and have an anonymous implementation of SQLiteTransactionListener here that calls the existing methods on the referenceDelegate.

I know this will result in some duplication here since we start transactions in two places, but I think it keeps the contract cleaner.  WDYT?

Reply:
Good point, though by moving it in here, it clutters the transaction logic, making it more difficult to audit. See 496e614eab3f70b576b64565e9d1ccdea6e8a4d1

However, we can create an adapter to (a) reduce the duplication and (b) move the clutter out from the transaction methods. See 2931f24093352674b4bca2664dd2c77c8d261914

wdyt?

Reply:
This probably isn't worth churning on too much, but having an "Adapter" class introduces some conceptual complexity.  I'd actually just have the listener be a private class member initialized via an anonymous class:

```
// Create a SQLiteTransactionListener that delegates to our referenceDelegate.
private final SQLiteTransactionListener transactionListener = 
  new SQLiteTransactionListener() {
          @Override
          public void onBegin() {
            referenceDelegate.onTransactionStarted();
          }
           @Override
          public void onCommit() {
            referenceDelegate.onTransactionCommitted();
          }
           @Override
          public void onRollback() {}
        });

...

db.beginTransactionWithListener(transactionListener);
```

Reply:
WDYT?  I'm fine with you just submitting as-is too. :-)

Reply:
Yeah, that makes more sense; done.